### ISSUE TITLE
error configmap  "aws-auth" does not exist and  creating IAM Policy ekstest-additional: ConcurrentModification: Found Tagris has different internal

### ISSUE THREAD

**Issue by jriveralivaque** [terraform-aws-modules/terraform-aws-eks#2525]
## Description
se esta implementando un eks con nodos administrados pero s iempre falla
por favor revisar o dar una solucion 

adjunto codigo y el error que sale 

```
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.default.token

  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_name]
  }
}

data "aws_caller_identity" "current" {}
data "aws_availability_zones" "available" {}

locals {
  name            = "ekstest"
  cluster_version = "1.25"
  region          = "eu-west-2"

  
  tags = {
    Cluster   = local.name
    Project = "culqi"
    PCI= "false"
  }
}

resource "aws_eks_addon" "coredns" {
  cluster_name      = module.eks.cluster_id
  addon_name        = "coredns"
  resolve_conflicts = "OVERWRITE"
  depends_on = [
    module.eks
  ]
}


################################################################################
# EKS Module
################################################################################

module "eks" {
  source = "../../module/eks"

  cluster_name                   = local.name
  cluster_version                = local.cluster_version
  cluster_enabled_log_types= [
    "api",
    "audit",
    "authenticator",
    "controllerManager",
    "scheduler"
  ]
  cluster_endpoint_public_access = true #double ccheck

  # IPV6
  cluster_ip_family = "ipv6"
  create_cni_ipv6_iam_policy = true


  vpc_id                   = data.aws_vpc.vpc_name-application.id
  subnet_ids               = ["${data.aws_subnet.subnets01.id}", "${data.aws_subnet.subnets02.id}", "${data.aws_subnet.subnets03.id}"]
  #control_plane_subnet_ids = module.vpc.intra_subnets

 


  eks_managed_node_group_defaults = {
    ami_type       = "AL2_x86_64"
    instance_types = ["m5.xlarge"]

    attach_cluster_primary_security_group = true
   
    iam_role_attach_cni_policy = true
  }
  cluster_addons = {
     
      kube-proxy = {
        most_recent = true
      }
      vpc-cni = {
        most_recent              = true
        before_compute           = true
        service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
        configuration_values = jsonencode({
          env = {
            # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
            ENABLE_PREFIX_DELEGATION = "true"
            WARM_PREFIX_TARGET       = "1"
          }
        })
      }
    }



  manage_aws_auth_configmap = true
  eks_managed_node_groups = {
 
    complete = {
      name            = "default_node"
      use_name_prefix = true
      

      subnet_ids = ["${data.aws_subnet.subnets01.id}", "${data.aws_subnet.subnets02.id}", "${data.aws_subnet.subnets03.id}"]
      use_custom_launch_template = false

      min_size     = 2
      max_size     = 10
      desired_size = 4

      capacity_type        = "SPOT"
      force_update_version = true
      instance_types       = ["m5.xlarge"]
      disk_size = 180
      labels = local.tags

      ami_id                     = data.aws_ami.eks_default.image_id
      enable_bootstrap_user_data = true

      pre_bootstrap_user_data = <<-EOT
        export FOO=bar
      EOT

      post_bootstrap_user_data = <<-EOT
        echo "you are free little kubelet!"
      EOT

      

      taints = [
        {
          key    = "dedicated"
          value  = "gpuGroup"
          effect = "NO_SCHEDULE"
        }
      ]

      update_config = {
        max_unavailable_percentage = 33 # or set `max_unavailable`
      }

      description = "EKS managed node group  launch template"

      ebs_optimized           = true
      disable_api_termination = false
      enable_monitoring       = true

      block_device_mappings = {
        xvda = {
          device_name = "/dev/xvda"
          ebs = {
            volume_size           = 180
            volume_type           = "gp3"
            iops                  = 3000
            throughput            = 150
            encrypted             = true
            kms_key_id            = module.ebs_kms_key.key_arn
            delete_on_termination = true
          }
        }
      }

      # Remote access cannot be specified with a launch template
      remote_access = {
        ec2_ssh_key               = module.key_pair.key_pair_name
        source_security_group_ids = [aws_security_group.remote_access.id]
      }

      metadata_options = {
        http_endpoint               = "enabled"
        http_tokens                 = "required"
        http_put_response_hop_limit = 2
        instance_metadata_tags      = "disabled"
      }

      create_iam_role          = true
      iam_role_name            = "eks-managed-node-group-complete-example"
      iam_role_use_name_prefix = false
      iam_role_description     = "EKS managed node group complete example role"
      iam_role_tags = {
        Purpose = "Protector of the kubelet"
      }
      iam_role_additional_policies = {
        AmazonEC2ContainerRegistryReadOnly = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
        additional                         = aws_iam_policy.node_additional.arn
      }

      tags = {
        ExtraTag = "EKS managed node group"
      }
    }
  }
  
# Extend cluster security group rules
  cluster_security_group_additional_rules = {
    ingress = {
      description                = "EKS Cluster allows 443 port to get API call"
      type                       = "ingress"
      from_port                  = 443
      to_port                    = 443
      protocol                   = "TCP"
      cidr_blocks                = ["0.0.0.0/0"]
      source_node_security_group = false
    }
  }
  tags = local.tags
}

################################################################################
# Supporting Resources
################################################################################

module "vpc_cni_irsa" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
  version = "~> 5.0"

  role_name_prefix      = "VPC-CNI-IRSA"
  attach_vpc_cni_policy = true
  vpc_cni_enable_ipv6   = true

  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }

  tags = local.tags
}

module "ebs_kms_key" {
  source  = "terraform-aws-modules/kms/aws"
  version = "~> 1.5"

  description = "Customer managed key to encrypt EKS managed node group volumes"

  # Policy
  key_administrators = [
    data.aws_caller_identity.current.arn
  ]

  key_service_roles_for_autoscaling = [
    # required for the ASG to manage encrypted volumes for nodes
    "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
    # required for the cluster / persistentvolume-controller to create encrypted PVCs
    module.eks.cluster_iam_role_arn,
  ]

  # Aliases
  aliases = ["eks/${local.name}/ebs"]

  tags = local.tags
}

module "key_pair" {
  source  = "terraform-aws-modules/key-pair/aws"
  version = "~> 2.0"

  key_name_prefix    = local.name
  create_private_key = true

  tags = local.tags
}

resource "aws_security_group" "remote_access" {
  name_prefix = "${local.name}-remote-access"
  description = "Allow remote SSH access"
  vpc_id      = data.aws_vpc.vpc_name-application.id

  ingress {
    description = "SSH access"
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["10.0.0.0/8"]
  }

   egress {
    from_port        = 0
    to_port          = 0
    protocol         = "-1"
    cidr_blocks      = ["0.0.0.0/0"]
    ipv6_cidr_blocks = ["::/0"]
  }

  tags = merge(local.tags, { Name = "${local.name}-remote" })
}



resource "aws_iam_policy" "node_additional" {
  name        = "${local.name}-additional"
  description = "node additional policy"

  policy = jsonencode({
    Version = "2012-10-17"
    Statement = [
      {
        Action = [
          "ec2:Describe*",
        ]
        Effect   = "Allow"
        Resource = "*"
      },
    ]
  })

  tags = local.tags
}

data "aws_ami" "eks_default" {
  most_recent = true
  owners      = ["amazon"]

  filter {
    name   = "name"
    values = ["amazon-eks-node-${local.cluster_version}-v*"]
  }
}



data "aws_vpc" "vpc_name-application" {
  filter {
    name   = "tag:Name"
    values = ["${var.vpc_name-application}"]
  }
}


data "aws_subnet" "subnets01" {
   filter {
     name   = "tag:Name"
     values = ["vpc-application-private-us-west-2b"]
   }
}

data "aws_subnet" "subnets02" {
   filter {
     name   = "tag:Name"
     values = ["vpc-application-private-us-west-2a"]
   }
}

data "aws_subnet" "subnets03" {
   filter {
     name   = "tag:Name"
     values = ["vpc-application-private-us-west-2c"]
   }
}

data "aws_eks_cluster_auth" "default" {
  name = local.name
}

resource "aws_iam_policy" "additional" {
  name        = "${local.name}-additional"
  description = "Example usage of node additional policy"

  policy = jsonencode({
    Version = "2012-10-17"
    Statement = [
      {
        Action = [
          "ec2:Describe*",
        ]
        Effect   = "Allow"
        Resource = "*"
      },
    ]
  })

  tags = local.tags
}
```

- [ ] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: 
terraform-aws-eks 19.10.1

- Terraform version: 
Terraform v1.3.9

### Terminal Output Screenshot(s)

![2023-03-17 21_40_25-main tf - eks - Visual Studio Code](https://user-images.githubusercontent.com/42852702/226080187-52070c59-4337-4e59-a577-8e0115f4329b.png)

**Comment by BarnumD** [comment]
Make sure you have something like this in your top level code, alongside your eks module. Than it should work.

```
data "aws_eks_cluster" "eks" {
  name = module.eks.cluster_name
}

data "aws_eks_cluster_auth" "eks" {
  name = module.eks.cluster_name
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.eks.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.eks.token
}
```

**Comment by dhruvinmaniar123** [comment]
but th

> I had the same problem and to work around it, I enabled the `create_aws_auth_configmap = true` flag, which will return another informing that the aws-auth already exists. After that, I commented out the `create_aws_auth_configmap = true` flag, leaving only the `manage_aws_auth_configmap = true`, and it bypasses the error.

yes that's the workaround I am currently using too but need a permanent fix as I can't do this frequently everytime when creating a new one.

**Comment by stevenpitts** [comment]
Un-stale

**Comment by Dhirva** [comment]
Use this provider 
```
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command     = "aws"
    args        = ["eks", "get-token", "--cluster-name", module.eks.cluster_name]
  }
}
```

**Comment by sansmoraxz** [comment]
> Use this provider it works
> 
> ```
> provider "kubernetes" {
>   host                   = module.eks.cluster_endpoint
>   cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
>   exec {
>     api_version = "client.authentication.k8s.io/v1beta1"
>     command     = "aws"
>     args        = ["eks", "get-token", "--cluster-name", module.eks.cluster_name]
>   }
> }
> ```

Not while you run destroy if you got dependent modules.

Alternatively you can try using eksctl command as a `null_resource`

https://eksctl.io/usage/iam-identity-mappings/

**Comment by bryantbiggs** [comment]
> Not while you run destroy if you got dependent modules.

Why not?

**Comment by sansmoraxz** [comment]
Just what I observed. It tries to dial the connection after the cluster was deleted. Some helm charts were pending which we had to manually remove from state file.
Maybe we approached it wrong, and should have specified the dependencies better.

However what I observed is it's better to extract that to a local first and pass those on to the provider.

**Comment by bryantbiggs** [comment]
if you want to share your terraform configs I can take a look - it sounds like something else is amiss

**Comment by sansmoraxz** [comment]
@bryantbiggs you may be right about it being something else.
Can't seem to find that configuration though, or be able to replicate it.

**Comment by sansmoraxz** [comment]
Btw for anyone facing issues with the original issue and want to do it in a single pass, assuming you have eksctl available you may try, like below:

```terraform
resource "null_resource" "eksctl_add_auth_configmap" {
  triggers = {
    cluster_name = module.eks.cluster_name
    region       = var.region
    cluster_endpoint = module.eks.cluster_endpoint
    karpernter_role_arn = module.karpenter.role_arn
  }

  provisioner "local-exec" {
    command = <<EOT
           eksctl create iamidentitymapping \
             --cluster ${self.triggers.cluster_name} \
             --arn ${self.triggers.karpernter_role_arn} \
             --group system:bootstrappers \
             --group system:nodes \
             --username karpenter \
             --region ${self.triggers.region}
        EOT
    on_failure = fail
  }

  provisioner "local-exec" {
    when = destroy
    command = <<EOT
           eksctl delete iamidentitymapping \
             --cluster ${self.triggers.cluster_name} \
             --arn ${self.triggers.karpernter_role_arn} \
             --region ${self.triggers.region}
        EOT
    on_failure = fail
  }

  depends_on = [
    module.karpenter,
    module.eks,
  ]
}
```

**Comment by sansmoraxz** [comment]
For anyone having eksctl, can try something like below:

```terraform
resource "null_resource" "eksctl_add_auth_configmap" {
  triggers = {
    cluster_name = module.eks.cluster_name
    region       = var.region
    cluster_endpoint = module.eks.cluster_endpoint
    karpernter_role_arn = module.karpenter.role_arn
  }

  provisioner "local-exec" {
    command = <<EOT
           eksctl create iamidentitymapping \
             --cluster ${self.triggers.cluster_name} \
             --arn ${self.triggers.karpernter_role_arn} \
             --group system:bootstrappers \
             --group system:nodes \
             --username system:node:{{EC2PrivateDNSName}} \
             --region ${self.triggers.region}
        EOT
    on_failure = fail
  }

  provisioner "local-exec" {
    when = destroy
    command = <<EOT
           eksctl delete iamidentitymapping \
             --cluster ${self.triggers.cluster_name} \
             --arn ${self.triggers.karpernter_role_arn} \
             --region ${self.triggers.region}
        EOT
    on_failure = fail
  }

  depends_on = [
    module.karpenter,
    module.eks,
  ]
}
```
^ This one uses the [karpenter](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest/submodules/karpenter) module with [eks](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest).

**Comment by dev-travelex** [comment]
I have updated the provider section as mentioned above. But I still get this error:
```

"msg": "\nError: The configmap \"aws-auth\" does not exist\n\n  with module.eks.kubernetes_config_map_v1_data.aws_auth[0],\n  on .terraform/modules/eks/main.tf line 553, in resource \"kubernetes_config_map_v1_data\" \"aws_auth\":\n 553: resource \"kubernetes_config_map_v1_data\" \"aws_auth\" {",
    "rc": 1,
    "stderr": "\nError: The configmap \"aws-auth\" does not exist\n\n  with module.eks.kubernetes_config_map_v1_data.aws_auth[0],\n  on .terraform/modules/eks/main.tf line 553, in resource \"kubernetes_config_map_v1_data\" \"aws_auth\":\n 553: resource \"kubernetes_config_map_v1_data\" \"aws_auth\" {\n\n",
    "stderr_lines": [
        "",
        "Error: The configmap \"aws-auth\" does not exist",
        "",
        "  with module.eks.kubernetes_config_map_v1_data.aws_auth[0],",
        "  on .terraform/modules/eks/main.tf line 553, in resource \"kubernetes_config_map_v1_data\" \"aws_auth\":",
        " 553: resource \"kubernetes_config_map_v1_data\" \"aws_auth\" {",
        ""
```

If I also add `create_aws_auth_configmap = true` , I get this 
```

   "msg": "\nError: Post \"https://xxxxxxxxxxxxx.gr7.eu-west-1.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps\": dial tcp 100.68.84.210:443: i/o timeout\n\n  with module.eks.kubernetes_config_map.aws_auth[0],\n  on .terraform/modules/eks/main.tf line 536, in resource \"kubernetes_config_map\" \"aws_auth\":\n 536: resource \"kubernetes_config_map\" \"aws_auth\" {",
    "rc": 1,
```

**Comment by KouriR** [comment]
In my case I had to update my kubernetes provider to the following:
```
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)

  exec {
    api_version = var.kubeconfig_api_version
    command     = "aws"
    // This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--region", var.region, "--cluster-name", module.eks.cluster_name, "--output", "json"]
  }
}
```
This is needed because the `eks get-token` output doesn't default to json since 2.10.4.
https://github.com/aws/aws-cli/issues/7719#issuecomment-1458939693

**Comment by kostyanius** [comment]
> Make sure you have something like this in your top level code, alongside your eks module. Than it should work.
> 
> ```
> data "aws_eks_cluster" "eks" {
>   name = module.eks.cluster_name
> }
> 
> data "aws_eks_cluster_auth" "eks" {
>   name = module.eks.cluster_name
> }
> 
> provider "kubernetes" {
>   host                   = data.aws_eks_cluster.eks.endpoint
>   cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
>   token                  = data.aws_eks_cluster_auth.eks.token
> }
> ```

This works only if you already have created cluster already and you only need to add or adjust the aws0auth configmap, but it you will use run this from scratch before cluster creation then it throws an error like this:  `reading EKS Cluster (dev): couldn't find resource`

**Comment by ArnArm** [comment]
> > Make sure you have something like this in your top level code, alongside your eks module. Than it should work.
> > ```
> > data "aws_eks_cluster" "eks" {
> >   name = module.eks.cluster_name
> > }
> > 
> > data "aws_eks_cluster_auth" "eks" {
> >   name = module.eks.cluster_name
> > }
> > 
> > provider "kubernetes" {
> >   host                   = data.aws_eks_cluster.eks.endpoint
> >   cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
> >   token                  = data.aws_eks_cluster_auth.eks.token
> > }
> > ```
> 
> It works only if you have created cluster already and you only need to add or adjust the aws-auth configmap, but if you will use this from scratch before cluster creation then it throws an error like this: `reading EKS Cluster (dev): couldn't find resource` :(

We are facing exactly the same issue. Any updates?

**Comment by nmofonseca** [comment]
> Make sure you have something like this in your top level code, alongside your eks module. Than it should work.
> 
> ```
> data "aws_eks_cluster" "eks" {
>   name = module.eks.cluster_name
> }
> 
> data "aws_eks_cluster_auth" "eks" {
>   name = module.eks.cluster_name
> }
> 
> provider "kubernetes" {
>   host                   = data.aws_eks_cluster.eks.endpoint
>   cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
>   token                  = data.aws_eks_cluster_auth.eks.token
> }
> ```

I was facing the same issue, changed the Kubernetes provider config to use the data sources to retrieve the token instead of the  AWS cli, like I did before and has been mentioned by @BarnumD 

However this is really an awkward one, I was able to test the module several times with the approach of using ***aws cli*** several times then suddenly started getting the error out of the blue. There must be something about the Kubernetes provider and the exec option that is not working properly and not specifically to this module.  Regardless the flag is set or not , not sure what is the scenario we need the flag to create the config map but the config map is always created by EKS service, that's how he adds the creator of the cluster as admin, otherwise we wouldn't be able to do anything with the cluster afterwards, at least that is my understanding.

I am using the following versions:
aws cli: 2.9.19
Terraform 1.5.4
EKS module: 19.16.0
Kubernetes provider: v2.16.1
AWS provider: v4.67.0
kubectl provider: v1.14.0

I hope this helps

**Comment by Vermyndax** [comment]
We are also experiencing this issue.

**Comment by kostyanius** [comment]
Managed to solve it somehow. 
![зображення](https://github.com/terraform-aws-modules/terraform-aws-eks/assets/9063908/a35b416f-c6fa-4f6f-84eb-bbc942f3432c)
Here is my main.tf file:
```
terraform {
  cloud {
    organization = "YOUR ORGANIZATION HERE"
    workspaces {
      name = "test"
    }
  }
}


provider "aws" {
  region = var.aws_region
}


module "vpc" {
  source = "terraform-aws-modules/vpc/aws"

  name = "${var.env}-vpc"
  cidr = "${var.vpc_cidr_block}"

  azs             = ["${var.az1}", "${var.az2}", "${var.az3}"]
  private_subnets = ["${var.private_subnet_cidr_blocks[0]}", "${var.private_subnet_cidr_blocks[1]}", "${var.private_subnet_cidr_blocks[2]}"]
  public_subnets  = ["${var.public_subnet_cidr_blocks[0]}", "${var.public_subnet_cidr_blocks[1]}", "${var.public_subnet_cidr_blocks[2]}"]

#  enable_vpn_gateway = true

   enable_nat_gateway = true
   single_nat_gateway = true
   one_nat_gateway_per_az = false

  tags = {
    Terraform = "true"
    Environment = "${var.env}"
  }
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 19.15.3"

  cluster_name    = "${var.env}"
  cluster_version = "1.27"

  cluster_endpoint_public_access  = true

  cluster_addons = {
    coredns = {
      most_recent = true
    }
    kube-proxy = {
      most_recent = true
    }
    vpc-cni = {
      most_recent = true
    }
  }


  vpc_id                   = "${module.vpc.vpc_id}"
  subnet_ids               = ["${element(module.vpc.private_subnets, 0)}", "${element(module.vpc.private_subnets, 1)}", "${element(module.vpc.private_subnets, 2)}"]
  control_plane_subnet_ids = ["${element(module.vpc.public_subnets, 0)}", "${element(module.vpc.public_subnets, 1)}", "${element(module.vpc.public_subnets, 2)}"]

  # Self Managed Node Group(s)
  self_managed_node_group_defaults = {
    instance_type                          = "${var.instance_types[0]}"
    update_launch_template_default_version = true
    iam_role_additional_policies = {
      AmazonSSMManagedInstanceCore = "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore"
    }
  }

  self_managed_node_groups = {
    one = {
      name         = "mixed-1"
      max_size     = 5
      desired_size = 2

      use_mixed_instances_policy = true
      mixed_instances_policy = {
        instances_distribution = {
          on_demand_base_capacity                  = 0
          on_demand_percentage_above_base_capacity = 10
          spot_allocation_strategy                 = "capacity-optimized"
        }

        override = [
          {
            instance_type     = "${var.instance_types[1]}"
            weighted_capacity = "1"
          },
          {
            instance_type     = "${var.instance_types[0]}"
            weighted_capacity = "2"
          },
        ]
      }
    }
  }

  # EKS Managed Node Group(s)
  eks_managed_node_group_defaults = {
    instance_types =  ["${var.instance_types[0]}", "${var.instance_types[1]}", "${var.instance_types[2]}"]
  }

  eks_managed_node_groups = {
    blue = {}
    green = {
      min_size     = 1
      max_size     = 10
      desired_size = 1

      instance_types = ["${var.instance_types[0]}"]
      capacity_type  = "SPOT"
    }
  }

  # Fargate Profile(s)
  fargate_profiles = {
    default = {
      name = "default"
      selectors = [
        {
          namespace = "default"
        }
      ]
    }
  }

  # aws-auth configmap
  manage_aws_auth_configmap = true

  aws_auth_roles = [
    {
      rolearn  = "arn:aws:iam::${var.aws_account}:role/${var.aws_role}"
      username = "${var.aws_role}"
      groups   = ["system:masters"]
    },
  ]

  aws_auth_users = [
    {
      userarn  = "arn:aws:iam::${var.aws_account}:user/${var.aws_user}"
      username = "${var.aws_user}"
      groups   = ["system:masters"]
    },
    {
      userarn  = "arn:aws:iam::${var.aws_account}:root"
      username = "root"
      groups   = ["system:masters"]
    }
  ]

  aws_auth_accounts = [
    "${var.aws_account}"
  ]

  tags = {
    Environment = "${var.env}"
    Terraform   = "true"
  }
}

data "aws_eks_cluster" "this" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

data "aws_eks_cluster_auth" "this" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.this.endpoint
  token                  = data.aws_eks_cluster_auth.this.token
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.this.certificate_authority.0.data)
}
```
So, the last block is really important! Please keep attention on it and add dependency on node group creation. 

```
data "aws_eks_cluster" "this" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

data "aws_eks_cluster_auth" "this" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.this.endpoint
  token                  = data.aws_eks_cluster_auth.this.token
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.this.certificate_authority.0.data)
}
```
Hope it will save your time and help to solve an issue!

**Comment by bryantbiggs** [comment]
why this:
```hcl
  vpc_id                   = "${module.vpc.vpc_id}"
  subnet_ids               = ["${element(module.vpc.private_subnets, 0)}", "${element(module.vpc.private_subnets, 1)}", "${element(module.vpc.private_subnets, 2)}"]
  control_plane_subnet_ids = ["${element(module.vpc.public_subnets, 0)}", "${element(module.vpc.public_subnets, 1)}", "${element(module.vpc.public_subnets, 2)}"]
```

instead of:
```hcl
  vpc_id                   = module.vpc.vpc_id
  subnet_ids               = module.vpc.private_subnets
  control_plane_subnet_ids = module.vpc.public_subnets
```

**Comment by kostyanius** [comment]
It was just my initiative of replacement for the subnets. In theory it should not affect the plan execution. 
So, you can to ignore it and just to add the "data sources to you code.

**Comment by nmofonseca** [comment]
hello,

@kostyanius thanks so much for sharing, I went on the route to move the aws-auth managed outside of the manage, just tested your approach this morning and it seems to be working for me too.

**Comment by benjamin-awd** [comment]
I didn't have much success with the above methods -- I ended up using the kubectl provider to apply the `aws_auth_configmap` as a manifest:

```
locals {
  users = [
    for user in data.aws_iam_group.admins.users :
    {
      userarn  = user.arn
      username = user.user_name
      groups = [
        "system:masters"
      ]
    }
  ]

aws_auth_configmap = {
    apiVersion = "v1"
    kind = "ConfigMap"
    metadata = {
      name = "aws-auth"
      namespace = "kube-system"
    }
    data = {
      mapUsers = yamlencode(local.users)
    }
  }
}

provider "kubectl" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.default.token
  load_config_file       = false
}

resource "kubectl_manifest" "aws_auth_configmap" {
  yaml_body = yamlencode(local.aws_auth_configmap)

  force_conflicts = true 
}
```

Theoretically you could use the `kubernetes_manifest` resource if you don't want to touch the kubectl provider, but unfortunately it requires access to the cluster API during the planning phase which may or may not work for your use case.

**Comment by nmofonseca** [comment]
> ```terraform
> aws_eks_cluster_auth
> ```

I understand, I went that route to but I since reverted back @kostyanius  work-around.

Have you made both data sources for aws_eks_cluster_auth and aws_eks_cluster_auth dependent of the eks module node groups?

my kubernetes provider config looks like this now:

 ```hcl
data "aws_eks_cluster" "cluster" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

data "aws_eks_cluster_auth" "clusterauth" {
  name = module.eks.cluster_name
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.cluster.endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.clusterauth.token
}

```

**Comment by benjamin-awd** [comment]
> Have you made both data sources for aws_eks_cluster_auth and aws_eks_cluster_auth dependent of the eks module node groups?


Yeah, gave that a try but unfortunately didn't work.

**Comment by kostyanius** [comment]
@benjamin-awd 
I have 3 possible solution, but so as i am not forced anymore to reproduce the issue so it is only a theory. You need to try it and tell us whether it solved your issue or not.
1. Update your aws-cli. I have seen that in some of the cases people claimed on aws-uath error when they had obsolate aws-cli version. I am using aws-cli/2.9.22. What is yours?
2. Try to update the module on the new version AWS EKS Module on 19.16.0 Published August 4, 2023 
3. true to add something like `create_aws_auth_configmap = true`
> Determines whether to create the aws-auth configmap. NOTE - this is  only intended for scenarios where the configmap does not exist (i.e. -  when using only self-managed node groups). Most users should use manage_aws_auth_configmap
And if all the variants haven`t helped to solve the issue then please hsare your module confugurtion here and order we could try it as well.

**Comment by ptimofee** [comment]
None of the provided workarounds fully work for me.
The one provided by @kostyanius works only for the initial deployment.
Later if I want to change some parameters (for example, disk size within node group) for the that cluster I get error 
```
Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 127.0.0.1:80: connect: connection refused
```
My config:
manage_aws_auth_configmap = true
create_aws_auth_configmap = false

**Comment by kostyanius** [comment]
@ptimofee 

  // Note: `disk_size`, and `remote_access` can only be set when using the EKS managed node group default launch template
  // This module defaults to providing a custom launch template to allow for custom security groups, tag propagation, etc.
  // use_custom_launch_template = false
  // disk_size = 50
  
  https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest/submodules/eks-managed-node-group

**Comment by ptimofee** [comment]
@kostyanius I didn't say "disk_size" exactly.
In my case it's volume_size within block_device_mappings in a node group.

**Comment by mrijaz** [comment]
finally got it fixed for me
make sure below 3 lines are exactly same

cluster_endpoint_public_access  = true
subnet_ids               = module.vpc.private_subnets
control_plane_subnet_ids = module.vpc.public_subnets

also,
manage_aws_auth_configmap = true
---------------
provider.tf

provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command     = "aws"
    args        = ["eks", "get-token", "--cluster-name", module.eks.cluster_name, "--profile", "default"]
  }
}

also pls check default profile in ~/.aws/credentials and ~/.aws/config is correct

**Comment by ClaytonOlleyNutrien** [comment]
Un-stale

**Comment by Almenon** [comment]
I tried the solutions in this issue and ran into all sorts of different errors. I ended up managing the aws_auth configmap with our own code instead of having the module handle it. It's ~50 lines of extra code, but it's well worth it. My provider looks like so:

```hcl
provider "kubernetes" {
  host                   = module.this_cluster.cluster_endpoint
  cluster_ca_certificate = base64decode(module.this_cluster.cluster_certificate_authority_data)
  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.this_cluster.cluster_name]
  }
}
```

**Comment by gmisura** [comment]
Pardon my (possible) ignorance, but should it be fairly easy for the code to use the var.aws_auth_roles merge/concat into the aws-auth configmap?

I mean, aws-auth configmap is created even when manage_aws_auth_configmap (and create_aws_auth_configmap) is(are) false, and I'm using an eks managed node group (the managed node group role is added to aws-auth configmap in the groups: system:bootstrappers, system:nodes

Seems like it should read/use the var: aws_auth_roles and merge/concat but doesn't (I tested with v19.20.0).

**Comment by antonbabenko** [comment]
This issue has been resolved in version 20.0.0 :tada:
